### PR TITLE
Vivek/stalwart 181

### PR DESCRIPTION
**Pull request by VivekSharma2607**
### :nut_and_bolt: Description: What code changed, and why?
Created new API for counting the Assets named as CountAssets

### :+1: Definition of Done 
- API should throw error if start date > end date
- API should throw error if end date - start date > 90 days
- All Exceptions are Handled Properly
- Ensure logs have no unnecessary data.
- Test coverage for the new feature is done to at least 70%
- Swagger Documentation updated
- Smoke Test done.
- Ensure Build and Integration Pipelines are Green.
- PR has 2 approvers.
- All Code Review Comments are Resolved.

README doc should be updated, if needed.
### :athletic_shoe: How to Build and Test the Change
rebuild components/compliance-service
rebuild components/automate-gateway
curl for testing 
`curl --location --request POST 'https://a2-dev.test/api/v0/compliance/reporting/assets/count' \
--header 'api-token: GD2cbnOQTLICLPMhs7OiFZZfGtA=' \
--header 'Content-Type: application/json' \
--data-raw '{
    "filters": [
        {
            "type": "start_time",
            "values": [
                "2022-05-05T00:00:00Z"
            ]
        },
        {
            "type": "end_time",
            "values": [
                "2022-08-05T00:00:00Z"
            ]
        }
    ]
}'`
### :white_check_mark: Checklist

**All PRs** must tick these:

- [X] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [X ] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [X] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [X] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [X] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [X] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [X] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [ ] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

File: components/compliance-service/reporting/relaxting/assets.go [hunk 1]
```diff
@@ -272,3 +275,56 @@ func getSummaryAssetAggResult(aggRoot *elastic.SearchResult) *AssetSummary {
...
+
+	// get the total number of assets without any date range filters i.e all the assets present
+	boolquery := backend.getFiltersQueryForAssetFilters(filters)
+	totalAssets, err := backend.getAssets(ctx, boolquery)
+	if err != nil {
+		logrus.Errorf("The error while getting the Total Assests %v", err)
+		return nil, err
+	}
+	// getting the un-reachable assets as per the unreachable config
+	// Todo hardcoding the value for Reachable assets
```
Comment (dave-augustus) [inline components/compliance-service/reporting/relaxting/assets.go:288]: Remove the TODO line as it causes code smells

File: components/compliance-service/reporting/relaxting/assets.go [hunk 2]
```diff
@@ -272,3 +275,56 @@ func getSummaryAssetAggResult(aggRoot *elastic.SearchResult) *AssetSummary {
...
+	if err != nil {
+		logrus.Errorf("The error while getting the unreported Assests %v", err)
+		return nil, err
+	}
+	unreportedAsset := totalAssets - (unreachableAsset + reported)
+
+	// getting the collected assets as per the filters present
+	collectedAsset, err := backend.getCollectedAssets(ctx, reportedQuery)
+	if err != nil {
+		logrus.Errorf("The error while getting the unreported Assests %v", err)
```
Comment (dave-augustus) [inline components/compliance-service/reporting/relaxting/assets.go:310]:
```suggestion
		logrus.Errorf("The error while getting the collected assets: %v", err)
```

File: components/compliance-service/api/reporting/server/server.go [hunk 3]
```diff
@@ -914,3 +915,40 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
...
 	return nil*/
 }
+func (srv *Server)AssetCount(ctx context.Context , in *reporting.ListFilters) (*reporting.AssetSummary , error) {
+	formatttedFilters := formatFilters(in.Filters)
+	var count reporting.AssetSummary
+	if len(formatttedFilters["start_time"]) <= 0 || len(formatttedFilters["end_time"]) <= 0 {
+		return nil , &errorutils.InvalidError{}
+	} else if reflect.DeepEqual(formatttedFilters["start_time"] , formatttedFilters["end_time"]) {
+		return nil , &errorutils.InvalidError{}
+	} else {
```
Comment (dave-augustus) [inline components/compliance-service/api/reporting/server/server.go:925]: As previous conditions are returning, this else isn't required,  remove the else condition and put the block statemnt below line 925

File: components/compliance-service/api/reporting/server/server.go [hunk 4]
```diff
@@ -914,3 +915,40 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
 	}
 	return nil*/
 }
+func (srv *Server)AssetCount(ctx context.Context , in *reporting.ListFilters) (*reporting.AssetSummary , error) {
+	formatttedFilters := formatFilters(in.Filters)
+	var count reporting.AssetSummary
+	if len(formatttedFilters["start_time"]) <= 0 || len(formatttedFilters["end_time"]) <= 0 {
```
Comment (dave-augustus) [inline components/compliance-service/api/reporting/server/server.go:921]: log and return

File: components/compliance-service/api/reporting/server/server.go [hunk 5]
```diff
@@ -914,3 +915,40 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
...
+		return nil , &errorutils.InvalidError{}
+	} else if reflect.DeepEqual(formatttedFilters["start_time"] , formatttedFilters["end_time"]) {
+		return nil , &errorutils.InvalidError{}
```
Comment (dave-augustus) [inline components/compliance-service/api/reporting/server/server.go:924]: log and return

File: components/compliance-service/api/reporting/server/server.go [hunk 6]
```diff
@@ -914,3 +915,40 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
...
+		end_time,err := time.Parse(layout, end_date[0]) 
+		diff := int(end_time.Sub(start_time).Hours()/24)
+		if err != nil {
+			return nil , err
+		}else if diff > 90{
+			return nil , errors.Errorf("The Date Range should in range of 90 Days")
+		} else {
+			formatttedFilters, err := filterByProjects(ctx , formatttedFilters)
+			if err != nil {
+				return nil , errorutils.FormatErrorMsg(err , "")
```
Comment (dave-augustus) [inline components/compliance-service/api/reporting/server/server.go:939]: log and return

File: components/compliance-service/api/reporting/server/server.go [hunk 7]
```diff
@@ -914,3 +915,40 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
...
+			}
+			assets, err := srv.es.GetSummary(ctx, formatttedFilters)
+			if err != nil {
+				return nil , errorutils.FormatErrorMsg(err , "")
```
Comment (dave-augustus) [inline components/compliance-service/api/reporting/server/server.go:943]: log and return

**Review by dave-augustus — CHANGES_REQUESTED** [review]

File: components/compliance-service/api/reporting/server/server.go [hunk 8]
```diff
@@ -914,3 +915,40 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
...
 }
+func (srv *Server)AssetCount(ctx context.Context , in *reporting.ListFilters) (*reporting.AssetSummary , error) {
+	formatttedFilters := formatFilters(in.Filters)
+	var count reporting.AssetSummary
+	if len(formatttedFilters["start_time"]) <= 0 || len(formatttedFilters["end_time"]) <= 0 {
+		return nil , &errorutils.InvalidError{}
+	} else if reflect.DeepEqual(formatttedFilters["start_time"] , formatttedFilters["end_time"]) {
+		return nil , &errorutils.InvalidError{}
+	} else {
+		start_date := formatttedFilters["start_time"]
```
Comment (YashviJain01) [inline components/compliance-service/api/reporting/server/server.go:926]: validate this in another function. Rename the variables as well. Create a function which validated start time and end time. Take the layout as ttime.RFC3339.

File: components/compliance-service/api/reporting/server/server.go [hunk 9]
```diff
@@ -914,3 +915,40 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
...
+			formatttedFilters, err := filterByProjects(ctx , formatttedFilters)
+			if err != nil {
+				return nil , errorutils.FormatErrorMsg(err , "")
+			}
+			assets, err := srv.es.GetSummary(ctx, formatttedFilters)
+			if err != nil {
+				return nil , errorutils.FormatErrorMsg(err , "")
+			}
+			count.TotalAssets = assets.TotalAssets
+			count.Collected = assets.Collected
```
Comment (YashviJain01) [inline components/compliance-service/api/reporting/server/server.go:946]: Why are we taking two variables, can we do this in one?

File: components/compliance-service/api/reporting/server/server.go [hunk 10]
```diff
@@ -914,3 +915,48 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
...
+	}
+	assets, err = srv.es.GetSummary(ctx, formattedFilters)
+	if err != nil {
+		return nil , errorutils.FormatErrorMsg(err , "")
+	}
+	return assets , nil
+	
+}
+
+func Validate(startTime string , endTime string) (bool , error) {
```
Comment (YashviJain01) [inline components/compliance-service/api/reporting/server/server.go:950]: rename the function to what?

File: components/compliance-service/api/reporting/server/server.go [hunk 11]
```diff
@@ -914,3 +915,48 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
...
 	return nil*/
 }
+func (srv *Server)AssetCount(ctx context.Context , in *reporting.ListFilters) (*reporting.AssetSummary , error) {
+	formattedFilters := formatFilters(in.Filters)
+	var assets *reporting.AssetSummary
+	if len(formattedFilters["start_time"]) <= 0 || len(formattedFilters["end_time"]) <=0 {
+		logrus.Errorf("The startTime or endTime cannot be null")
+		return nil , &errorutils.InvalidError{}
+	} 
+	if reflect.DeepEqual(formattedFilters["start_time"] , formattedFilters["end_time"]) {
```
Comment (YashviJain01) [inline components/compliance-service/api/reporting/server/server.go:925]: add this logic in the function itself.

File: components/compliance-service/api/reporting/server/server.go [hunk 12]
```diff
@@ -914,3 +915,48 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
...
+		logrus.Errorf("The startTime or endTime cannot be equal")
+		return nil , &errorutils.InvalidError{}
+	}
+	timeRange, err := Validate(formattedFilters["start_time"][0] , formattedFilters["end_time"][0])
```
Comment (YashviJain01) [inline components/compliance-service/api/reporting/server/server.go:929]: this logic as well.

File: components/compliance-service/api/reporting/server/server.go [hunk 13]
```diff
@@ -914,3 +915,48 @@ func (srv *Server) GetReportContent(ctx context.Context, in *reporting.ReportCon
 	}
 	return nil*/
 }
+func (srv *Server)AssetCount(ctx context.Context , in *reporting.ListFilters) (*reporting.AssetSummary , error) {
+	formattedFilters := formatFilters(in.Filters)
+	var assets *reporting.AssetSummary
+	if len(formattedFilters["start_time"]) <= 0 || len(formattedFilters["end_time"]) <=0 {
```
Comment (YashviJain01) [inline components/compliance-service/api/reporting/server/server.go:921]: Add this logic in the function in validate function.

**Review by daveaugustus — APPROVED** [review]

**Pull request merged**